pip/index.py

```diff
@@ -669,10 +667,9 @@ def _sort_links(self, links):
         return no_eggs + eggs
 
     def _package_versions(self, links, search):
-        for link in self._sort_links(links):
-            v = self._link_package_versions(link, search)
-            if v is not None:
-                yield v
+        vs = [self._link_package_versions(link, search)
+              for link in self._sort_links(links)]
+        return [v for v in vs if v is not None]
```

Comment:
Of the two, I prefer a decorator or a local accumulator. Don't have a big preference between those two though.